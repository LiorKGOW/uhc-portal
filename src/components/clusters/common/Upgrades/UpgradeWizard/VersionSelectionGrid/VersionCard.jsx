import React from 'react';
import PropTypes from 'prop-types';
import {
  Card,
  CardTitle,
  CardBody,
  CardFooter,
  Label,
  Split,
  SplitItem,
  Button,
  Tooltip,
} from '@patternfly/react-core';
import { StarIcon } from '@patternfly/react-icons/dist/esm/icons/star-icon';
import { InfoCircleIcon } from '@patternfly/react-icons/dist/esm/icons/info-circle-icon';
import { versionRegEx } from '../../../../../../common/versionComparator';
import ExternalLink from '../../../../../common/ExternalLink';

const getReleaseNotesLink = (version) => {
  const { groups: parts } = versionRegEx.exec(version);
  const { major, minor, revision, patch } = parts;
  if (major !== '4' || patch !== undefined) {
    return null;
  }
  const docURL = `https://docs.openshift.com/container-platform/${major}.${minor}/release_notes/ocp-${major}-${minor}-release-notes.html#ocp-${major}-${minor}-${revision}`;
  return (
    <ExternalLink href={docURL} className="pf-c-button pf-m-link card-footer-button">
      View release notes
    </ExternalLink>
  );
};

const VersionAcknowledgementPopover = (version) => (
  <Tooltip
    content={`All clusters require an administrator acknowledgement before updating to OpenShift Container Platform ${version}`}
  >
    <Button variant="link" isInline aria-label="more information">
      <InfoCircleIcon />
    </Button>
  </Tooltip>
);

const VersionCard = (props) => {
  const {
    isSelected,
    isRecommended,
    version,
    onKeyDown,
    onClick,
    children,
    getUnMetClusterAcknowledgements,
  } = props;

  return (
    <Card
      className="version-card"
      id={version}
      onKeyDown={onKeyDown}
      onClick={onClick}
      isSelectableRaised
      isCompact
      isSelected={isSelected}
    >
      <CardTitle>
        <Split>
          <SplitItem>
            {version}{' '}
            {getUnMetClusterAcknowledgements(version).length > 0
              ? VersionAcknowledgementPopover(version)
              : null}
          </SplitItem>
          <SplitItem isFilled />
          {isRecommended && (
            <SplitItem>
              <Label color="blue" icon={<StarIcon />}>
                Recommended
              </Label>
            </SplitItem>
          )}
        </Split>
      </CardTitle>
      {children && <CardBody>{children}</CardBody>}
      <CardFooter>{getReleaseNotesLink(version)}</CardFooter>
    </Card>
  );
};

VersionCard.propTypes = {
  isSelected: PropTypes.bool,
  isRecommended: PropTypes.bool,
  version: PropTypes.string.isRequired,
  onKeyDown: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
  children: PropTypes.node,
  getUnMetClusterAcknowledgements: PropTypes.func,
};

VersionCard.defaultProps = {
  isRecommended: false,
};

export default VersionCard;
