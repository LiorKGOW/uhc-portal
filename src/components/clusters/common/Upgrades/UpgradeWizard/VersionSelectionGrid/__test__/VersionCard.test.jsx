import React from 'react';
import { shallow } from 'enzyme';
import { Tooltip } from '@patternfly/react-core';
import VersionCard from '../VersionCard';

let wrapper;
let onKeyDown;
let onClick;

describe('<VersionCard>', () => {
  beforeEach(() => {
    onKeyDown = jest.fn();
    onClick = jest.fn();
    wrapper = shallow(
      <VersionCard
        isRecommended
        version="4.5.20"
        onKeyDown={onKeyDown}
        onClick={onClick}
        getUnMetClusterAcknowledgements={() => []}
      >
        The latest on your current minor version.
      </VersionCard>,
    );
  });

  it('should render correctly when recommended', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when selected & not recommended', () => {
    wrapper.setProps(
      {
        isSelected: true,
        version: '4.5.16',
        isRecommended: false,
      },
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render a tooltip without any unmet cluster acknowledgements', () => {
    expect(wrapper.find(Tooltip)).toHaveLength(0);
  });

  it('should render tooltip when has unmet cluster acknowledgements', () => {
    wrapper.setProps(
      {
        getUnMetClusterAcknowledgements: () => [{ id: 'someUpgradeGateId' }],
      },
    );

    expect(wrapper).toMatchSnapshot();
  });
});
