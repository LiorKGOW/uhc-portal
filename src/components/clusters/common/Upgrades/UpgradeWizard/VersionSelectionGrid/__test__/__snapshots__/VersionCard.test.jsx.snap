// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VersionCard> should render correctly when recommended 1`] = `
<Card
  className="version-card"
  id="4.5.20"
  isCompact={true}
  isSelectableRaised={true}
  onClick={[MockFunction]}
  onKeyDown={[MockFunction]}
>
  <CardTitle>
    <Split>
      <SplitItem>
        4.5.20
         
      </SplitItem>
      <SplitItem
        isFilled={true}
      />
      <SplitItem>
        <Label
          color="blue"
          icon={<StarIcon />}
        >
          Recommended
        </Label>
      </SplitItem>
    </Split>
  </CardTitle>
  <CardBody>
    The latest on your current minor version.
  </CardBody>
  <CardFooter>
    <Button
      className="card-footer-button"
      variant="link"
    >
      <ExternalLink
        href="https://docs.openshift.com/container-platform/4.5/release_notes/ocp-4-5-release-notes.html#ocp-4-5-20"
      >
        View release notes
      </ExternalLink>
    </Button>
  </CardFooter>
</Card>
`;

exports[`<VersionCard> should render correctly when selected & not recommended 1`] = `
<Card
  className="version-card"
  id="4.5.16"
  isCompact={true}
  isSelectableRaised={true}
  isSelected={true}
  onClick={[MockFunction]}
  onKeyDown={[MockFunction]}
>
  <CardTitle>
    <Split>
      <SplitItem>
        4.5.16
         
      </SplitItem>
      <SplitItem
        isFilled={true}
      />
    </Split>
  </CardTitle>
  <CardBody>
    The latest on your current minor version.
  </CardBody>
  <CardFooter>
    <Button
      className="card-footer-button"
      variant="link"
    >
      <ExternalLink
        href="https://docs.openshift.com/container-platform/4.5/release_notes/ocp-4-5-release-notes.html#ocp-4-5-16"
      >
        View release notes
      </ExternalLink>
    </Button>
  </CardFooter>
</Card>
`;

exports[`<VersionCard> should render tooltip when has unmet cluster acknowledgements 1`] = `
<Card
  className="version-card"
  id="4.5.20"
  isCompact={true}
  isSelectableRaised={true}
  onClick={[MockFunction]}
  onKeyDown={[MockFunction]}
>
  <CardTitle>
    <Split>
      <SplitItem>
        4.5.20
         
        <Tooltip
          content="All clusters require an administrator acknowledgement before updating to OpenShift Container Platform 4.5.20"
        >
          <Button
            isInline={true}
            variant="link"
          >
            <InfoCircleIcon />
          </Button>
        </Tooltip>
      </SplitItem>
      <SplitItem
        isFilled={true}
      />
      <SplitItem>
        <Label
          color="blue"
          icon={<StarIcon />}
        >
          Recommended
        </Label>
      </SplitItem>
    </Split>
  </CardTitle>
  <CardBody>
    The latest on your current minor version.
  </CardBody>
  <CardFooter>
    <Button
      className="card-footer-button"
      variant="link"
    >
      <ExternalLink
        href="https://docs.openshift.com/container-platform/4.5/release_notes/ocp-4-5-release-notes.html#ocp-4-5-20"
      >
        View release notes
      </ExternalLink>
    </Button>
  </CardFooter>
</Card>
`;
